{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: jkobject/scprint\n"
     ]
    }
   ],
   "source": [
    "! lamin load jkobject/scprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: jkobject/scprint\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln\n",
    "from scdataloader import DataModule, Preprocessor, utils\n",
    "from scdataloader.preprocess import additional_postprocess, additional_preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m the database (0.77.0) is ahead of your installed lamindb package (0.76.12)\n",
      "\u001b[92m→\u001b[0m please update lamindb: pip install \"lamindb>=0.77,<0.78\"\n",
      "\u001b[93m!\u001b[0m source schema has additional modules: {'ourprojects'}\n",
      "consider mounting these schema modules to not encounter errors\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Collection(uid='dMyEX3NTfKOEYXyMKDD7', version='2024-07-01 00:00:00 ', is_latest=True, name='cellxgene-census', hash='nI8Ag-HANeOpZOz-8CSn', visibility=1, created_by_id=1, transform_id=22, run_id=27, created_at=2024-07-16 12:14:38 UTC)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cxg = ln.Collection.using(instance=\"laminlabs/cellxgene\").filter(name='cellxgene-census').last()\n",
    "cxg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Artifact(uid='nnSyG0O54y0MJ58m2jYr', version='2024-07-01 00:00:00 ', is_latest=True, description='A single-cell and spatially-resolved atlas of human breast cancers', key='cell-census/2024-07-01/h5ads/9fddb063-056d-4202-8b8a-4b0ee531d3ce.h5ad', suffix='.h5ad', type='dataset', size=839092209, hash='bpBJEd5AOqxfFUNOxmVrrA', n_observations=100064, _hash_type='md5-n', _accessor='AnnData', visibility=1, _key_is_virtual=False, storage_id=2, transform_id=22, run_id=27, created_by_id=1, created_at=2024-07-12 12:34:09 UTC)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file = cxg.artifacts.filter(description__contains=\"spatial\").first()\n",
    "file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m the database (0.77.0) is ahead of your installed lamindb package (0.76.12)\n",
      "\u001b[92m→\u001b[0m please update lamindb: pip install \"lamindb>=0.77,<0.78\"\n",
      "\u001b[93m!\u001b[0m source schema has additional modules: {'ourprojects'}\n",
      "consider mounting these schema modules to not encounter errors\n",
      "\u001b[93m!\u001b[0m run input wasn't tracked, call `ln.track()` and re-run\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 100064 × 29067\n",
       "    obs: 'donor_id', 'percent_mito', 'nCount_RNA', 'nFeature_RNA', 'celltype_major', 'celltype_minor', 'celltype_subset', 'subtype', 'gene_module', 'calls', 'normal_cell_call', 'CNA_value', 'batch_run', 'multiplexed', 'cryo_state', 'development_stage_ontology_term_id', 'cancer_type', 'ER', 'PR', 'HER2_IHC', 'HER2_ISH', 'HER2_ISH_ratio', 'Ki67', 'subtype_by_IHC', 'treatment_status', 'treatment_details', 'assay_ontology_term_id', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'suspension_type', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'disease_ontology_term_id', 'grade', 'cell_type_ontology_term_id', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_Three-D', 'X_umap'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata = file.load()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping layers:  KeysView(Layers with keys: )\n",
      "checking raw counts\n",
      "removed 0 non primary cells, 100064 renamining\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/content/scDataLoader/scdataloader/preprocess.py:179: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs[\"nnz\"] = np.array(np.sum(adata.X != 0, axis=1).flatten())[0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 0 cells, 100064 renamining\n",
      "Removed 80 genes.\n",
      "validating\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/content/scDataLoader/scdataloader/preprocess.py:248: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  data_utils.validate(adata, organism=adata.obs.organism_ontology_term_id[0])\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n",
      "/usr/local/lib/python3.10/dist-packages/lamindb/_can_validate.py:113: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if isinstance(values[0], list):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "startin QC\n",
      "Seeing 36967 outliers (36.94% of total dataset):\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/content/scDataLoader/scdataloader/preprocess.py:351: ImplicitModificationWarning: Trying to modify index of attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs.index = [str(uuid4()) for _ in range(adata.shape[0])]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "preprocessor = Preprocessor(\n",
    "    do_postp=False, # takes too much RAM in collab to do PCA\n",
    "    # additional_postprocess=additional_postprocess,\n",
    "    additional_preprocess=additional_preprocess,\n",
    "    force_preprocess=True,\n",
    ")\n",
    "adata = preprocessor(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m no run & transform got linked, call `ln.track()` & re-run\n",
      "\u001b[93m!\u001b[0m no run & transform got linked, call `ln.track()` & re-run\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "repr_error": "Out of range float values are not JSON compliant: nan",
       "type": "dataframe"
      },
      "text/html": [
       "\n",
       "  <div id=\"df-163785dd-0a30-44c4-ac75-b323ee2788c5\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>version</th>\n",
       "      <th>is_latest</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>hash</th>\n",
       "      <th>reference</th>\n",
       "      <th>reference_type</th>\n",
       "      <th>visibility</th>\n",
       "      <th>transform_id</th>\n",
       "      <th>meta_artifact_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OqJqEd34D7oDpcBH0000</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>test dataset</td>\n",
       "      <td>None</td>\n",
       "      <td>o7h1KdEaJ-OTy3otUuhABQ</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-12-05 08:15:00.435044+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-163785dd-0a30-44c4-ac75-b323ee2788c5')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-163785dd-0a30-44c4-ac75-b323ee2788c5 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-163785dd-0a30-44c4-ac75-b323ee2788c5');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ],
      "text/plain": [
       "                     uid version  is_latest          name description  \\\n",
       "id                                                                      \n",
       "1   OqJqEd34D7oDpcBH0000    None       True  test dataset        None   \n",
       "\n",
       "                      hash reference reference_type  visibility transform_id  \\\n",
       "id                                                                             \n",
       "1   o7h1KdEaJ-OTy3otUuhABQ      None           None           1         None   \n",
       "\n",
       "   meta_artifact_id run_id                       created_at  created_by_id  \n",
       "id                                                                          \n",
       "1              None   None 2024-12-05 08:15:00.435044+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m run input wasn't tracked, call `ln.track()` and re-run\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Collection(uid='wmFtqol8XiNCg0HY0000', is_latest=True, name='test', hash='uV4sCcfOEfYQKH2UzuVsfA', visibility=1, created_by_id=1, created_at=2024-12-05 08:46:02 UTC)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# we make a collection from this one dataset but we could use multiple.\n",
    "# this is the point of a collection\n",
    "file = ln.Artifact.from_anndata(adata, description=\"test anndata\")\n",
    "file.save()\n",
    "col = ln.Collection(file, name=\"test\")\n",
    "col.save()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ln.Collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m no run & transform got linked, call `ln.track()` & re-run\n",
      "\u001b[93m!\u001b[0m run input wasn't tracked, call `ln.track()` and re-run\n",
      "won't do any check but we recommend to have your dataset coming from local storage\n",
      "100.0% are aligned\n",
      "{'x': tensor([[8., 1., 2.,  ..., 0., 0., 0.],\n",
      "        [2., 1., 1.,  ..., 0., 0., 0.],\n",
      "        [1., 1., 1.,  ..., 0., 0., 0.],\n",
      "        ...,\n",
      "        [1., 2., 1.,  ..., 0., 0., 0.],\n",
      "        [1., 1., 4.,  ..., 0., 0., 0.],\n",
      "        [1., 2., 1.,  ..., 0., 0., 0.]]), 'genes': tensor([[10119,  5691,  9158,  ..., 66566, 14554, 45837],\n",
      "        [16656,  7803, 10563,  ..., 38870, 37021, 34603],\n",
      "        [11932,  5999,  7888,  ..., 60141, 41023, 66992],\n",
      "        ...,\n",
      "        [ 5208,  4965, 14835,  ..., 36221, 59153, 19157],\n",
      "        [17348,  6694,  2568,  ..., 67598,  1346, 36390],\n",
      "        [ 5745,  1316, 17974,  ...,  2935, 68319, 24508]], dtype=torch.int32), 'class': tensor([[ 0,  3],\n",
      "        [ 0,  0],\n",
      "        [ 0,  2],\n",
      "        [ 0,  0],\n",
      "        [ 0, 20],\n",
      "        [ 0, 21],\n",
      "        [ 0, 10],\n",
      "        [ 0, 11],\n",
      "        [ 0,  1],\n",
      "        [ 0, 19],\n",
      "        [ 0,  8],\n",
      "        [ 0,  8],\n",
      "        [ 0, 13],\n",
      "        [ 0, 12],\n",
      "        [ 0,  4],\n",
      "        [ 0,  4],\n",
      "        [ 0,  8],\n",
      "        [ 0,  4],\n",
      "        [ 0,  1],\n",
      "        [ 0,  8],\n",
      "        [ 0,  0],\n",
      "        [ 0, 18],\n",
      "        [ 0, 19],\n",
      "        [ 0,  6],\n",
      "        [ 0,  6],\n",
      "        [ 0, 12],\n",
      "        [ 0,  4],\n",
      "        [ 0, 21],\n",
      "        [ 0, 19],\n",
      "        [ 0, 12],\n",
      "        [ 0, 11],\n",
      "        [ 0,  4],\n",
      "        [ 0, 18],\n",
      "        [ 0, 10],\n",
      "        [ 0, 13],\n",
      "        [ 0, 11],\n",
      "        [ 0,  7],\n",
      "        [ 0,  1],\n",
      "        [ 0,  8],\n",
      "        [ 0, 19],\n",
      "        [ 0,  1],\n",
      "        [ 0,  7],\n",
      "        [ 0,  8],\n",
      "        [ 0,  0],\n",
      "        [ 0,  8],\n",
      "        [ 0, 12],\n",
      "        [ 0, 18],\n",
      "        [ 0,  0],\n",
      "        [ 0, 20],\n",
      "        [ 0, 18],\n",
      "        [ 0, 20],\n",
      "        [ 0, 12],\n",
      "        [ 0, 21],\n",
      "        [ 0, 10],\n",
      "        [ 0,  9],\n",
      "        [ 0,  2],\n",
      "        [ 0,  6],\n",
      "        [ 0,  6],\n",
      "        [ 0,  6],\n",
      "        [ 0, 18],\n",
      "        [ 0,  4],\n",
      "        [ 0, 19],\n",
      "        [ 0,  2],\n",
      "        [ 0,  4]], dtype=torch.int32), 'tp': tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
      "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
      "        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), 'depth': tensor([ 1606.,  4594.,  3283.,  6009.,  4330.,  5094.,   752., 10750.,  2454.,\n",
      "         2550.,  2499.,   325.,  3215.,  1988.,  1045.,  1457.,  1649.,  5107.,\n",
      "         1838.,   910.,  9914.,   655.,   687.,  1940.,  1550., 19502.,  2394.,\n",
      "         4067.,  4307.,  7326.,  3110.,  2606.,  3007.,  6607.,  1100.,   686.,\n",
      "         1614., 14072.,  1494.,  2181., 12114.,  2598.,  1923.,  5877.,   923.,\n",
      "         3841.,  3008., 26831.,  7262.,  6660.,  4065., 12844., 11445.,  8375.,\n",
      "        10492.,  3238.,  3916.,  1331.,  3704.,  5035.,  1536.,  4224.,  3538.,\n",
      "         1971.])}\n"
     ]
    }
   ],
   "source": [
    "datamodule = DataModule(\n",
    "    collection_name=\"test\",\n",
    "    organisms=[\"NCBITaxon:9606\"],  # organism that we will work on\n",
    "    how=\"random expr\",  # for the collator (random set of expr genes will be selected)\n",
    "    max_len=300,  # only 300 genes will be shown\n",
    "    batch_size=64,\n",
    "    do_gene_pos=False,\n",
    "    num_workers=1,\n",
    "    use_default_col=True,\n",
    "    clss_to_weight=[\"organism_ontology_term_id\", \"cell_type_ontology_term_id\"],\n",
    "    all_clss=[\"organism_ontology_term_id\", \"cell_type_ontology_term_id\"],\n",
    "    hierarchical_clss=[\"cell_type_ontology_term_id\"],\n",
    "    validation_split=0.1,\n",
    "    test_split=0,\n",
    ")\n",
    "datamodule.setup()\n",
    "for i in datamodule.train_dataloader():\n",
    "    # pass #or do pass\n",
    "    print(i)\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now go have a look at the mapped.py file in [here](https://github.com/laminlabs/lamindb/blob/39e0f529a41d9cbc475e52fe3aa0b8095af21494/lamindb/core/_mapped_collection.py) and the `__getitem__()` function which can be modified to work with [spatialdata](https://spatialdata.scverse.org/en/latest/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now let's work on imputation of spatial transcriptomic data\n",
    "\n",
    "- we will download a model\n",
    "- we will use a denoiser tool (made for scPRINT)\n",
    "  - it takes a dataset\n",
    "  - removes a lot of counts\n",
    "  - uses scPRINT to recover/impute the removed counts\n",
    "  - compares the pre-removal with the prediction\n",
    "- we will make sure that we improve correlation to ground truth post denoising"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "! git clone https://github.com/mahmoodlab/HEST.git\n",
    "! pip install -e ./HEST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import urllib.request\n",
    "import torch\n",
    "\n",
    "from scprint import scPrint\n",
    "from scprint.tasks import Denoiser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ckpt_path = \"large.ckpt\"\n",
    "if not os.path.exists(ckpt_path):\n",
    "    url = \"https://huggingface.co/jkobject/scPRINT/resolve/main/large.ckpt\"\n",
    "    urllib.request.urlretrieve(url, ckpt_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RuntimeError caught: scPrint is not attached to a `Trainer`.\n"
     ]
    }
   ],
   "source": [
    "model = scPrint.load_from_checkpoint(\n",
    "        ckpt_path,\n",
    "        precpt_gene_emb=None,\n",
    "        # triton gets installed so it must think it has cuda enabled\n",
    "        transformer=\"fast\", #else normal, without flashattention\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ml4ig1/miniconda3/envs/scprint/lib/python3.10/site-packages/datatree/__init__.py:10: DeprecationWarning: The repository `xarray-contrib/datatree` has been archived! It is no longer maintained as of October 2024. Datatree functionality has been migrated upstream to `pydata/xarray`, so please use `xarray.DataTree` instead. To help you migrate a guide can be found by following https://github.com/pydata/xarray/discussions/9680\n",
      "  warnings.warn(\n",
      "/home/ml4ig1/miniconda3/envs/scprint/lib/python3.10/site-packages/numba/core/decorators.py:246: RuntimeWarning: nopython is set for njit and is ignored\n",
      "  warnings.warn('nopython is set for njit and is ignored', RuntimeWarning)\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'read' from 'spatialdata' (/home/ml4ig1/miniconda3/envs/scprint/lib/python3.10/site-packages/spatialdata/__init__.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mdatasets\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mhest\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mHESTData\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m load_hest\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mspatialdata\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m read\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mhuggingface_hub\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m login\n\u001b[1;32m      7\u001b[0m login(token\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhf_TIFDQNfqTNRABLRuEYfMpVkPnETjoVsscf\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'read' from 'spatialdata' (/home/ml4ig1/miniconda3/envs/scprint/lib/python3.10/site-packages/spatialdata/__init__.py)"
     ]
    }
   ],
   "source": [
    "import datasets\n",
    "from hest.HESTData import load_hest\n",
    "from spatialdata import read\n",
    "\n",
    "from huggingface_hub import login\n",
    "\n",
    "login(token=\"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\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d7c35cda2e7e4f3993ac79e2f60473f1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "hest.py:   0%|          | 0.00/2.17k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "781f2377f9204ea2ade741bd0a89d10e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "README.md:   0%|          | 0.00/6.42k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "76a6801605824577a24c6174141fe49f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Fetching 34 files:   0%|          | 0/34 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cbc96e77fa50468aaab6127e26ec016f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "metadata/TENX95.json:   0%|          | 0.00/2.17k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "320d47cbaf0b4532801b02e85f0e5ced",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_cellvit_seg.geojson.zip:   0%|          | 0.00/96.8M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6534ec0d0d7649f483f51d248c8a4ca0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "metadata/TENX99.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0adcf5348367494b9b575ac6bed7d9bb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99.h5:   0%|          | 0.00/3.18G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5c5325214a5842738908d832d0f2f1e1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95.h5:   0%|          | 0.00/1.18G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7f6b3a353b6d419382b68dabaaf607b8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_cellvit_seg.parquet:   0%|          | 0.00/166M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bf40d447cea14124a131bf5d6cd390a4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_cellvit_seg.geojson.zip:   0%|          | 0.00/153M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "52296984e23a4dcea2531bd735cc24d7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_cellvit_seg.parquet:   0%|          | 0.00/106M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d5d35463b7e24db2976a5c2b1d543a20",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_patch_vis.jpg:   0%|          | 0.00/719k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "74c76201253944729aa5e9fe0d8332f1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_patch_vis.jpg:   0%|          | 0.00/660k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c355a265403e49a982c7bed602d2fac1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_pixel_size_vis.png:   0%|          | 0.00/222k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dc33248cb1ed437c8e397fa1bd717ca4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_pixel_size_vis.png:   0%|          | 0.00/122k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ff52f57ca97f4c7caa8e6168c5b79b51",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_spatial_plots.png:   0%|          | 0.00/293k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "390de71925d647ca8bd9195b15401518",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_spatial_plots.png:   0%|          | 0.00/160k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a30ff56c52ac4baca61807313cc092ce",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99.h5ad:   0%|          | 0.00/30.7M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "337336d3a6bf43e89c09e3ff2846db09",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95.h5ad:   0%|          | 0.00/17.7M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "28f48b465fdb4f299d7aae807c496fb2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_downscaled_fullres.jpeg:   0%|          | 0.00/153k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "718048e83486449990a5f481febf32fd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_downscaled_fullres.jpeg:   0%|          | 0.00/60.1k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "509bb759f1f54d09a0669d00a67f415e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tissue_seg/TENX95_contours.geojson:   0%|          | 0.00/188k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2d19272f03e84bfc91028237431d43e8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_vis.jpg:   0%|          | 0.00/144k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "283724ca065c46ef865bf7f686656a06",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tissue_seg/TENX99_contours.geojson:   0%|          | 0.00/126k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b034a1f32a344b19bf20e0a2363cf58c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_vis.jpg:   0%|          | 0.00/236k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "69a93645abe5494db8a5135f3cf9c097",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_transcripts.parquet:   0%|          | 0.00/6.46G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "22794d4c68f148b980b1c62a54aff434",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_transcripts.parquet:   0%|          | 0.00/4.86G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "38d6313606ee4b3d88d9490881a2baaa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95.tif:   0%|          | 0.00/6.57G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "006c4eba38e748e88b12bbbb930fdec6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99.tif:   0%|          | 0.00/11.8G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bf3dac9dcc574508ba235e57384c18c8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_xenium_cell_seg.geojson.zip:   0%|          | 0.00/142M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "da287924a62846469fcf915639336222",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_xenium_cell_seg.parquet:   0%|          | 0.00/118M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "da1ed6d659ba4a50adbd2a4e1b4ce52a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_xenium_nucleus_seg.geojson.zip:   0%|          | 0.00/141M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bb6a444b8eec44d392eabdd5136deb47",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX95_xenium_nucleus_seg.parquet:   0%|          | 0.00/119M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b2c8120f1f58454aab4c6f37d155d0d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_xenium_cell_seg.geojson.zip:   0%|          | 0.00/65.7M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c6d9ce03987644018ea5dd098dd8358e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_xenium_cell_seg.parquet:   0%|          | 0.00/83.0M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c055d9f18d04428dbd9c5745f8f527e3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_xenium_nucleus_seg.geojson.zip:   0%|          | 0.00/65.2M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6dd23f550e1c4b8f918c75d74bdf909b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX99_xenium_nucleus_seg.parquet:   0%|          | 0.00/82.8M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping cell vit segmentation...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [00:07<00:00,  3.80s/it]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b732c7e14d134568b29f9d26b927fd1b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Generating train split: 0 examples [00:00, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "local_dir='../hest_data' # hest will be dowloaded to this folder\n",
    "\n",
    "ids_to_query = ['TENX95', 'TENX99'] # list of ids to query\n",
    "\n",
    "list_patterns = [f\"*{id}[_.]**\" for id in ids_to_query]\n",
    "dataset = datasets.load_dataset(\n",
    "    'MahmoodLab/hest', \n",
    "    cache_dir=local_dir,\n",
    "    patterns=list_patterns,\n",
    "    trust_remote_code=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DatasetDict({\n",
       "    train: Dataset({\n",
       "        features: ['path'],\n",
       "        num_rows: 2\n",
       "    })\n",
       "})"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/2 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ml4ig1/miniconda3/envs/scprint/lib/python3.10/site-packages/hestcore/wsi.py:27: UserWarning: CuImage is not available. Ensure you have a GPU and cucim installed to use GPU acceleration.\n",
      "  warnings.warn(\"CuImage is not available. Ensure you have a GPU and cucim installed to use GPU acceleration.\")\n",
      "100%|██████████| 2/2 [00:00<00:00,  2.10it/s]\n"
     ]
    }
   ],
   "source": [
    "hdatas = load_hest('../hest_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata2 = hdatas[1].to_spatial_data(fullres=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata1 = hdatas[0].to_spatial_data(fullres=True)\n",
    "sdata1.write(\"TENX95.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata1 = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata2 = hdatas[1].to_spatial_data(fullres=True)\n",
    "sdata2.write(\"TENX99.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11845 × 541\n",
       "    obs: 'in_tissue', 'pxl_col_in_fullres', 'pxl_row_in_fullres', 'array_col', 'array_row', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito', 'region', 'instance_id'\n",
       "    var: 'mito', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    uns: 'spatial', 'spatialdata_attrs'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sdata1.tables['table']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import squidpy as sq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_counts=10)\n",
    "sc.pp.filter_genes(adata, min_cells=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)\n",
    "sc.tl.leiden(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.4,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata, percent_top=(10, 20, 50, 150), inplace=True)\n",
    "cprobes = (\n",
    "    adata.obs[\"control_probe_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords = (\n",
    "    adata.obs[\"control_codeword_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "print(f\"Negative DNA probe count % : {cprobes}\")\n",
    "print(f\"Negative decoding count % : {cwords}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dn = Denoiser(\n",
    "    max_cells=1000, #number of cells which will be processed\n",
    "    batch_size=16,\n",
    "    num_workers=8,\n",
    "    max_len=2000, # we will work on 2000 genes (input and output)\n",
    "    downsample=0.7, #we are removing 70% of the counts,\n",
    "    # should be modified to make the data look more like st\n",
    "    predict_depth_mult=3, # how much to increase expression\n",
    "    dtype=torch.float32,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Collection(uid='4PJVEHXDgr489o5afYxr', is_latest=True, name='all', description='all files that I could keep', hash='fYW4GdioNfHJg10RSAd6', visibility=1, created_by_id=1, created_at=2024-03-27 09:57:23 UTC)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ln.Collection.filter(name=\"all\").one()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "np.sqrt(1024)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 63/63 [00:07<00:00,  7.96it/s]\n"
     ]
    }
   ],
   "source": [
    "metrics, random_indices, genes, expr_pred = dn(\n",
    "    model=model,\n",
    "    adata=adata,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scprint",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
